src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStack.cs

```diff
@@ -130,24 +161,35 @@ internal JsonConverter Initialize(JsonTypeInfo jsonTypeInfo, bool supportContinu
             return jsonTypeInfo.PropertyInfoForTypeInfo.ConverterBase;
         }
 
+        /// <summary>
+        /// Gets the nested JsonTypeInfo before resolving any polymorphic converters
+        /// </summary>
+        public JsonTypeInfo PeekNestedJsonTypeInfo()
+        {
+            Debug.Assert(Current.PolymorphicSerializationState != PolymorphicSerializationState.PolymorphicReEntryStarted);
+            return _count == 0 ? Current.JsonTypeInfo : Current.JsonPropertyInfo!.JsonTypeInfo;
+        }
+
         public void Push()
         {
             if (_continuationCount == 0)
             {
-                if (_count == 0)
+                Debug.Assert(Current.PolymorphicSerializationState != PolymorphicSerializationState.PolymorphicReEntrySuspended);
+
+                if (_count == 0 && Current.PolymorphicSerializationState == PolymorphicSerializationState.None)
                 {
-                    // Performance optimization: reuse the first stackframe on the first push operation.
-                    // NB need to be careful when making writes to Current _before_ the first `Push`
-                    // operation is performed.
+                    // Perf enhancement: do not create a new stackframe on the first push operation
+                    // unless the converter has primed the current frame for polymorphic dispatch.
                     _count = 1;
+                    _indexOffset = 1; // currentIndex := _count - 1;
```

Comment:
I might be missing something but isn't this the only place where you assign this? If so isn't this always one when `_count` is > 0?

Reply:
The offset is always 1, with the exception of serializations where the root value is polymorphic. This is a corner case our infrastructure is not capable of handling, so we need to change what the first `Push()` operation is doing in that case.